python/llm/src/ipex_llm/transformers/npu_models/lm_head.py

```diff
@@ -0,0 +1,150 @@
+        split_num: int = 2,
+        profile: bool = False,
+        device: str = "NPU",
+        dtype: np.dtype = np.int8,
+    ):
+        """Initialize the QLinear class.
+
+        Args:
+            inC (int): input channels
+            outC (int): output channels
+            batch (int): batch
+            profile (bool): Enable/Disable profiling. Defaults to False.
+            device (str): Target device, default to "NPU".
+            dtype (np.dtype): weights datatype. Defaults to np.int8.
+
+        """
+        super().__init__(profile, device)
+        self.inC, self.outC = inC, outC
+        self.batch = batch
+
+        input = self.parameter((self.batch, self.inC))
+
+        self.split_num = split_num
+        split_size = self.inC // split_num // 2 * 2
+
+        for i in range(self.split_num):
+            start_idx = i * split_size
+            end_idx = (i + 1) * split_size if i < self.split_num - 1 else self.inC
+            input_slice = self.slice(input, begin=[0, start_idx],
+                                     end=[self.batch, end_idx])
+            linear_slice = self.linear(input_slice, outC, split_size, bias=False, wt_dtype=dtype)
+            if i == 0:
+                res = linear_slice
+            else:
+                res += linear_slice
+
+        print("start compiling lm_head")
+        self.compile()
+        print("end compiling lm_head")
+
+    def run(
+        self, X: np.ndarray
+    ) -> np.ndarray:
+        """Run the layer:  $X * (W * S)^T$ .
+
+        Args:
+            X (np.ndarray): activation
+
+        Raises:
+            RuntimeError: Input, weights or scale shape mismatch
+
+        Returns:
+            np.ndarray: result
+        """
+        self.prefetchWeights(1, verify_size=False)
+        self.set_input_tensor(X, 0)
+        self.elapsed = backend_lib.run(self._mm)
+        if len(self.out) == 1:
+            return self.out[0]
+        return self.out
+
+
+class SlicedLMHead(nn.Module):
+    def __init__(self, weight, bias, split_num):
+        super().__init__()
+        self.split_num = split_num
+        self.outC, self.inC = weight.shape
+        split_size = weight.size(1) // split_num // 2 * 2
+        self.lm_heads = nn.Sequential()
+        for i in range(split_num):
+            new_linear = torch.nn.Linear(0, 0, bias=False)
+            start_idx = i * split_size
+            end_idx = (i + 1) * split_size if i < split_num - 1 else weight.size(1)
+            new_weight = torch.nn.Parameter(weight[:, start_idx:end_idx],
+                                            requires_grad=False)
+            new_linear.weight = new_weight
+            new_linear.in_features = new_weight.size(1)
+            new_linear.out_features = new_weight.size(0)
+            self.lm_heads.append(new_linear)
+        self.bias = bias
+
+    def forward(self, hidden_states):
+        if hidden_states.size(0) * hidden_states.size(1) == 1 and hasattr(self, "fused_lm_head"):
```

Comment:
can we assume that `fused_lm_head` is always available?

Reply:
Yes we can, if we don't forget to call  `model.lm_head.get_fused_lm_head()` in `optimize_llm` function. I will remove this check first.